### PR TITLE
Add Protobuf serialization for `torch.Tensor` 

### PR DESCRIPTION
**Pull request by karlhigley**
This PR also extracts a Torch serde from the msgpack serialization and reuses it in the Protobuf serialization.

Depends on OpenMined/syft-proto#21.

### CODE REVIEW

File: syft/serde/protobuf/torch_serde.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+def _unbufferize_torch_tensor(
+    worker: AbstractWorker, protobuf_tensor: "TorchTensorPB"
+) -> torch.Tensor:
+    """
+    This function converts a serialized torch tensor into a torch tensor
+    using pickle.
+
+    Args:
+        tensor_tuple (bin): serialized obj of torch tensor. It's a tuple where
```
Comment (vvmnnnkv) [inline syft/serde/protobuf/torch_serde.py]: Outdated comment

File: syft/serde/protobuf/torch_serde.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    return deserializer(worker, tensor_bin)
+
+
+def protobuf_tensor_serializer(worker: AbstractWorker, tensor: torch.Tensor) -> TensorDataPB:
+    """Strategy to serialize a tensor using Protobuf"""
+    dtype = TORCH_DTYPE_STR[tensor.dtype]
+
+    protobuf_tensor = TensorDataPB()
+
+    if tensor.is_quantized:
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply syft/serde/protobuf/torch_serde.py]: Looks that way. I think the `torch_tensor_serializer` ought to still work though. Might be a good first issue to fix those.

File: syft/serde/protobuf/torch_serde.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        return torch.tensor(data, dtype=dtype).reshape(size)
+
+
+# Bufferize/Unbufferize Torch Tensors
+
+
+def _bufferize_torch_tensor(worker: AbstractWorker, tensor: torch.Tensor) -> bin:
+    """
+    This function converts a torch tensor into a serliaized torch tensor
+    using pickle. We choose to use this because PyTorch has a custom and
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply syft/serde/protobuf/torch_serde.py]: It's true as far as it goes, but could be clearer and more complete. Depending on the worker's serializer, `TorchTensorPB` can contain the binary output of either the Torch or Numpy serializers, or a "generic" Protobuf representation for cross-platform communication. I'll update it.

File: syft/serde/protobuf/torch_serde.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    if hasattr(tensor, "child"):
+        chain = syft.serde.protobuf.serde._bufferize(worker, tensor.child)
+
+    protobuf_tensor = TorchTensorPB()
+    protobuf_tensor.id.CopyFrom(syft.serde.protobuf.serde.create_protobuf_id(tensor.id))
+    protobuf_tensor.shape.dims.extend(tensor.shape)
+
+    protobuf_tensor.serializer = SERIALIZERS_SYFT_TO_PROTOBUF[worker.serializer]
+    if worker.serializer == TENSOR_SERIALIZATION.ALL:
+        protobuf_tensor.contents_data.CopyFrom(serialized_tensor)
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply syft/serde/protobuf/torch_serde.py]: I haven't checked. It's awkward to assign directly, given the multiple serialization strategy pattern, but might be possible. What performance benchmark tools do we already have available?

**Comment by karlhigley** [comment]
Looks the notebook tests are failing because the notebooks can't import `syft.serde.torch.serde`, which is contained in this PR. I'm puzzled why they're trying to import code from this PR if they can't find code from this PR. Point me in the right direction on troubleshooting notebook tests?

**Pull request merged**